File: ts/src/resources/workflows.ts
import { HttpTransport } from "../http.js"

export class WorkflowsResource {
  /**
   * @param channelBase  - channel-service base URL (gateway/channel-service)
   * @param platformBase - platform base URL (gateway/platform)
   */
  constructor(
    private readonly http: HttpTransport,
    private readonly channelBase: string,
    private readonly platformBase: string,
  ) {}

  private get chV1() { return `${this.channelBase}/v1` }
  private get plV1() { return `${this.platformBase}/v1` }

  // ─── Channel automation (channel-service) ────────────────────────────────────

  async listChannelAutomation(params?: { channelType?: string }) {
    const url = new URL(`${this.chV1}/workflows/channel_automation`)
    if (params?.channelType) url.searchParams.set("channelType", params.channelType)
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  // ─── n8n workflows (platform) ────────────────────────────────────────────────

  async list(params?: Record<string, string>) {
    const url = new URL(`${this.plV1}/workflows`)
    if (params) Object.entries(params).forEach(([k, v]) => url.searchParams.set(k, v))
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  async get(workflowId: string) {
    return this.http.getFetch()(`${this.plV1}/n8n/workflows/${workflowId}`, { method: "GET" }).then(r => r.json())
  }

  async create(body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.plV1}/n8n/workflows`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async update(workflowId: string, body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.plV1}/n8n/workflows/${workflowId}`, {
      method: "PATCH",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async delete(workflowId: string): Promise<void> {
    await this.http.getFetch()(`${this.plV1}/n8n/workflows/${workflowId}`, { method: "DELETE" })
  }

  async getNew() {
    return this.http.getFetch()(`${this.plV1}/n8n/workflows/new`, { method: "GET" }).then(r => r.json())
  }

  // ─── Credentials ─────────────────────────────────────────────────────────────

  async listCredentials() {
    return this.http.getFetch()(`${this.plV1}/credentials`, { method: "GET" }).then(r => r.json())
  }

  async getCredential(credentialId: string) {
    const url = new URL(`${this.plV1}/n8n/credentials/${credentialId}`)
    url.searchParams.set("includeData", "true")
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  async createCredential(body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.plV1}/n8n/credentials`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async updateCredential(credentialId: string, body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.plV1}/n8n/credentials/${credentialId}`, {
      method: "PATCH",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async deleteCredential(credentialId: string): Promise<void> {
    await this.http.getFetch()(`${this.plV1}/n8n/credentials/${credentialId}`, { method: "DELETE" })
  }

  // ─── Channel credentials (channel-service) ───────────────────────────────────

  async getChannelCredential(credentialId: string) {
    return this.http.getFetch()(`${this.chV1}/channels/credentials/${credentialId}`, { method: "GET" }).then(r => r.json())
  }

  async updateChannelCredential(credentialId: string, body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.chV1}/channels/credentials/${credentialId}`, {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async deleteChannelCredential(credentialId: string): Promise<void> {
    await this.http.getFetch()(`${this.chV1}/channels/credentials/${credentialId}`, { method: "DELETE" })
  }
}

File: py/src/imbrace/resources/workflows.py
from typing import Any, Dict, Optional
from ..http import HttpTransport, AsyncHttpTransport


class WorkflowsResource:
    """Workflows domain — Sync.

    @param channel_base - channel-service base URL (gateway/channel-service)
    @param platform_base - platform service base URL (gateway/platform)
    """

    def __init__(self, http: HttpTransport, channel_base: str, platform_base: str):
        self._http = http
        self._ch = channel_base.rstrip("/")
        self._pl = platform_base.rstrip("/")

    @property
    def _ch_v1(self) -> str:
        return f"{self._ch}/v1"

    @property
    def _pl_v1(self) -> str:
        return f"{self._pl}/v1"

    # ─── Channel-service workflows ───────────────────────────────────────────────

    def list(self, tag: Optional[str] = None) -> Dict[str, Any]:
        params = {}
        if tag:
            params["tag"] = tag
        return self._http.request("GET", f"{self._ch_v1}/automations", params=params).json()

    def list_channel_automation(self, channel_type: Optional[str] = None) -> Dict[str, Any]:
        params = {}
        if channel_type:
            params["channelType"] = channel_type
        return self._http.request("GET", f"{self._ch_v1}/automations/channel_automation",
                                  params=params).json()

    def create(self, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._ch_v1}/automations", json=body).json()

    def update(self, workflow_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("PATCH", f"{self._ch_v1}/automations/{workflow_id}", json=body).json()

    # ─── Platform n8n workflows ──────────────────────────────────────────────────

    def list_n8n(self) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._pl_v1}/workflows").json()

    def get_n8n(self, id: str) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._pl_v1}/n8n/workflows/{id}").json()

    def create_n8n(self, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._pl_v1}/n8n/workflows", json=body).json()

    def update_n8n(self, id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("PATCH", f"{self._pl_v1}/n8n/workflows/{id}", json=body).json()

    def delete_n8n(self, id: str) -> None:
        self._http.request("DELETE", f"{self._pl_v1}/n8n/workflows/{id}")


class AsyncWorkflowsResource:
    """Workflows domain — Async."""

    def __init__(self, http: AsyncHttpTransport, channel_base: str, platform_base: str):
        self._http = http
        self._ch = channel_base.rstrip("/")
        self._pl = platform_base.rstrip("/")

    @property
    def _ch_v1(self) -> str:
        return f"{self._ch}/v1"

    @property
    def _pl_v1(self) -> str:
        return f"{self._pl}/v1"

    async def list(self, tag: Optional[str] = None) -> Dict[str, Any]:
        params = {}
        if tag:
            params["tag"] = tag
        res = await self._http.request("GET", f"{self._ch_v1}/automations", params=params)
        return res.json()

    async def list_channel_automation(self, channel_type: Optional[str] = None) -> Dict[str, Any]:
        params = {}
        if channel_type:
            params["channelType"] = channel_type
        res = await self._http.request("GET", f"{self._ch_v1}/automations/channel_automation",
                                       params=params)
        return res.json()

    async def create(self, body: Dict[str, Any]) -> Dict[str, Any]:
        res = await self._http.request("POST", f"{self._ch_v1}/automations", json=body)
        return res.json()

    async def update(self, workflow_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        res = await self._http.request("PATCH", f"{self._ch_v1}/automations/{workflow_id}", json=body)
        return res.json()

    async def list_n8n(self) -> Dict[str, Any]:
        res = await self._http.request("GET", f"{self._pl_v1}/workflows")
        return res.json()

    async def get_n8n(self, id: str) -> Dict[str, Any]:
        res = await self._http.request("GET", f"{self._pl_v1}/n8n/workflows/{id}")
        return res.json()

    async def create_n8n(self, body: Dict[str, Any]) -> Dict[str, Any]:
        res = await self._http.request("POST", f"{self._pl_v1}/n8n/workflows", json=body)
        return res.json()
